File: playground/Dart-Doom-Fire/fire.dart
import 'dart:html';
import 'dart:math' as Math;
import 'dart:async';
import 'dart:convert';


//Variaveis globais
var firePixelArray = [];
var fireColorsPalette = [['7,7,7'],['31,7,7'],['47,15,7'],['71,15,7'],['87,23,7'],['103,31,7'],['119,31,7'],['143,39,7'],['159,47,7'],['175,63,7'],['191,71,7'],['199,71,7'],['223,79,7'],['223,87,7'],['223,87,7'],['215,95,7'],['215,95,7'],['215,103,15'],['207,111,15'],['207,119,15'],['207,127,15'],['207,135,23'],['199,135,23'],['199,143,23'],['199,151,31'],['191,159,31'],['191,159,31'],['191,167,39'],['191,167,39'],['191,175,47'],['183,175,47'],['183,183,47'],['183,183,55'],['207,207,111'],['223,223,159'],['239,239,199'],['255,255,255']];
int fireWidth = 50;
int fireHeight = 50;
var debug = false;

//Funcao que inicializa o codigo
void main(){

    createFireDataStructure();
    createFireSource();
    renderFire();

    runfire();    
} 

void createFireDataStructure(){
    
    //multiplicando a altura pela largura para criar a tabela
    var numberOfPixels = fireHeight * fireWidth;

    //preenche com 0 - "potencia" do fogo
      for (int i = 0; i < numberOfPixels; i++) {
        firePixelArray.add(0);
      }
}

void createFireSource() {
    for (var column = 0; column < fireWidth; column++) {
        /*
            Como o for utiliza base 0 e o array base 1, ao criar a variavel overflowPixelIndex, ele trará um numero a mais
            Sabendo disso, o pixelIndex faz o calculo de forma que se subtraia uma linha, e incremente uma coluna.
            Ex:
            Array de 2 x 2 = 4 posições, ou de 0 a 3
            
            0 1
            2 3
            
            overflowPixelIndex = 2 x 2 = 4
            * para a primeira rodada do for
            column = 0
            pixelIndex = 4 - 2 + 0 = 2
            * Segunda rodada
            column = 1
            pixelIndex = 4 - 2 + 1 = 3
        */
        var overflowPixelIndex = fireWidth * fireHeight;
        var pixelIndex = (overflowPixelIndex - fireWidth) + column;

        firePixelArray[pixelIndex] = 36;
        
    }
    
}

void renderFire() {
    
    //monta a tabela no hmtl para criar o fogo
    var html = '<table cellpadding=0 cellspacing=0>';

    //percorre as linhas
    for (var row = 0; row < fireHeight; row++) {
        
        html += '<tr>';
        
        //percorre as colunas
        for (var column = 0; column < fireWidth; column++) {
            //encontra o indice
            var pixelIndex = column + (fireWidth * row);
            //grava o array de intensidade
            var fireIntensity = firePixelArray[pixelIndex];
            var color = fireColorsPalette[fireIntensity].toString();
            color = color.substring(1,(color.length)-1);

            
            if (debug == true) {
                html += '<td>';
                html +=  '<div class="pixel-index">${pixelIndex} </div>';  
                html += '<div style="color: rgb(${color})">${fireIntensity}</div>';
                html += '</td>';
            }else{

                html += '<td class="pixel" style="background-color: rgb(${color})"> </td>';
            }


        }
        
        html += '</tr>';    
        }
        html += '</table>';
    //atribui o valor a tag
    DivElement fireCanvas = querySelector('#fireCanvas');
    fireCanvas.setInnerHtml(html,treeSanitizer: new NullTreeSanitizer());

}

void calculateFirePropagation(){
    /*
        Para cada pixel percorrido, ele olhará o de baixo e subtrai 1
        Causando o efeito de fogo
    */
    for (var column = 0; column < fireWidth; column++) {
        for (var row = 0; row < fireHeight; row++) {
            //encontra o indice
            var pixelIndex = column + (fireWidth * row);

            updateFireIntensityPerPixel(pixelIndex);           
        }
    }

    renderFire();
}

void updateFireIntensityPerPixel(currentPixelIndex) {
    /*
        Pega o pixel atual e soma a largura, assim posiciona no pixel a baixo.
        Se o indice for maior que o tamanho do array, não faz nada
    */
    var belowPixelIndex = currentPixelIndex + fireWidth;

    if (belowPixelIndex >= fireWidth * fireHeight) {
        return;
    }

    var random = new Math.Random();
    var decay =  random.nextInt(3);
    
    var belowPixelIntensity = firePixelArray[belowPixelIndex];
    var newFireIntensity = 
        belowPixelIntensity - decay >= 0 ? belowPixelIntensity - decay : 0;

    
    firePixelArray[currentPixelIndex - decay] = newFireIntensity;
}



void setDebug(){
    /*
        Ativa ou desativa o modo de debug
    */
    
    if (debug == false) {
        fireWidth = 35;
        fireHeight = 25;
        debug = true;
    } else {
        fireWidth = 42;
        fireHeight = 42;
        debug = false;
    }

    createFireDataStructure();
    //console.log(firePixelArray);
    createFireSource();
}

void runfire() {
  const oneSec = const Duration(milliseconds:10);
  new Timer.periodic(oneSec, (Timer t) => calculateFirePropagation());
}


class NullTreeSanitizer implements NodeTreeSanitizer {
  /*
    Burlando a segurança do dart
  */
void sanitizeTree(Node node) {}
}

File: playground/controllers-rgb-and-scale-in-pure-javascript/fire.js

const firePixelsArray = []
let fireWidth = 60
let fireHeight = 40
let debug = false
var sentido = 0
const fireColorsPalette = [];

/*
const fireColorsPalette = [
{"b":7,"g":7,"r":7},
{"b":31,"g":7,"r":7},
{"b":47,"g":15,"r":7},
{"b":71,"g":15,"r":7},
{"b":87,"g":23,"r":7},
{"b":103,"g":31,"r":7},
{"b":119,"g":31,"r":7},
{"b":143,"g":39,"r":7},
{"b":159,"g":47,"r":7},
{"b":175,"g":63,"r":7},
{"b":191,"g":71,"r":7},
{"b":199,"g":71,"r":7},
{"b":223,"g":79,"r":7},
{"b":223,"g":87,"r":7},
{"b":223,"g":87,"r":7},
{"b":215,"g":95,"r":7},
{"b":215,"g":95,"r":7},
{"b":215,"g":103,"r":15},
{"b":207,"g":111,"r":15},
{"b":207,"g":119,"r":15},
{"b":207,"g":127,"r":15},
{"b":207,"g":135,"r":23},
{"b":199,"g":135,"r":23},
{"b":199,"g":143,"r":23},
{"b":199,"g":151,"r":31},
{"b":191,"g":159,"r":31},
{"b":191,"g":159,"r":31},
{"b":191,"g":167,"r":39},
{"b":191,"g":167,"r":39},
{"b":191,"g":175,"r":47},
{"b":183,"g":175,"r":47},
{"b":183,"g":183,"r":47},
{"b":183,"g":183,"r":55},
{"b":207,"g":207,"r":111},
{"b":223,"g":223,"r":159},
{"b":239,"g":239,"r":199},
{"b":255,"g":255,"r":255}]
*/


function changeRGB(r, g, b, contador = 9){

    //contador = 9;
    
    for(i = 0; i< 37; i++){

        if( (r * i) > 255){
            r = 0
        }
        if( (g * i) > 255){
            g = 0
        }
        if( (b * i) > 255){
            b = 0
        }

        fireColorsPalette[i] = {'r':r * contador,'g':g * contador,'b':b * contador};
        console.log(fireColorsPalette)
    }

    r++
    g++
    b++

}

function start(){

    changeRGB(5,19,37, 47);
    createFireDataStructure()
    createFireSource()

  setInterval(calculateFirePropagation, 50)
}

function createFireDataStructure(){
    const numberOfPixels = fireWidth * fireHeight

    for (let i = 0; i < numberOfPixels; i++) {
        firePixelsArray[i] = 0
      }
}

function calculateFirePropagation(){
    for (let column = 0; column < fireWidth; column++) {
        for (let row = 0; row < fireHeight; row++) {
            const pixelIndex = column + ( fireWidth * row )
            //console.log(column);
            updateFireIntensityPerPixel(pixelIndex)
        }
      }

      renderFire()
}

function changeWindDirection(value){
    sentido = value
}

function updateFireIntensityPerPixel(currentPixelIndex){

    const belowPixelIndex = currentPixelIndex + fireWidth
    if(belowPixelIndex >= fireWidth * fireHeight){
        return
    }

    const decay = Math.floor(Math.random() * 3)
    const belowPixelFireIntensity = firePixelsArray[belowPixelIndex]
    const newFireIntensity = 
        belowPixelFireIntensity - decay >= 0 ? belowPixelFireIntensity - decay : 0

        switch(sentido){
            case 0: firePixelsArray[currentPixelIndex - decay] = newFireIntensity; break;
            case 1: firePixelsArray[currentPixelIndex] = newFireIntensity; break;
            case 2: firePixelsArray[currentPixelIndex + decay] = newFireIntensity; break;
        }
}

function renderFire(){
    let html = '<table cellpadding=0>'

    for(let row = 0; row < fireHeight; row++){
        html += '<tr>'

        for(let column = 0; column < fireWidth; column++){
            const pixelIndex = column + ( fireWidth * row )
            const fireIntensity = firePixelsArray[pixelIndex]
            const color = fireColorsPalette[fireIntensity]
            const colorString = `${color.r},${color.g},${color.b}`

            if(debug === true){
                html += '<td>'
                html += `<div class="pixel-index">${pixelIndex}</div>`
                html += `<div style="color: rgb(${colorString})">${fireIntensity}</div>`
                html += '</td>'
            }else{

                html += `<td class="pixel" style="background-color: rgb(${colorString})"`
                html += '</td>'
            }
        }

        html += '</tr>'
    }

    html += '</table>'

    document.querySelector('#fireCanvas').innerHTML = html
}

function createFireSource() {
  for (let column = 0; column <= fireWidth; column++) {
    const overflowPixelIndex = fireWidth * fireHeight
    const pixelIndex = (overflowPixelIndex - fireWidth) + column

    firePixelsArray[pixelIndex] = 36
  }
}

function destroyFireSource(){
    for(let column = 0; column <= fireWidth; column++){
        console.log(fireHeight);
        const overflowPixelIndex = fireWidth * fireHeight
        const pixelIndex = (overflowPixelIndex - fireWidth) + column

        firePixelsArray[pixelIndex] = 0
    }
}


function increaseFireSource(){
    for(let column = 0; column <= fireWidth; column++){
        const overflowPixelIndex = fireWidth * fireHeight
        const pixelIndex = (overflowPixelIndex - fireWidth) + column
        const currentFireIntensity = firePixelsArray[pixelIndex]
    
        if(currentFireIntensity < 36){
            const increase = Math.floor(Math.random() * 14)
            const newFireIntensity = 
                currentFireIntensity + increase >= 36 ? 36 : currentFireIntensity + increase

                firePixelsArray[pixelIndex] = newFireIntensity
        }
    }
}

function decreaseFireSource() {
    for (let column = 0; column <= fireWidth; column++) {
      const overflowPixelIndex = fireWidth * fireHeight
      const pixelIndex = (overflowPixelIndex - fireWidth) + column
      const currentFireIntensity = firePixelsArray[pixelIndex]
  
      if (currentFireIntensity > 0) {
        const decay = Math.floor(Math.random() * 14)
        const newFireIntensity =
          currentFireIntensity - decay >= 0 ? currentFireIntensity - decay : 0
  
          firePixelsArray[pixelIndex] = newFireIntensity
      }
    }
  }


function toggleDebugMode(){
    if (debug === false) {
        fireWidth = 25
        fireHeight = 17
        debug = true
      } else {
        fireWidth = 60
        fireHeight = 40
        debug = false
      }

    createFireDataStructure()
    createFireSource()
}

start()